internal/github/contributions.go

```diff
@@ -20,53 +25,133 @@ type Contributions struct {
 	Days               []Day
 }
 
-func (g *GithubService) GetContributionsByUsername(ctx context.Context, username string) (*Contributions, error) {
-	var contributionsQuery struct {
-		User struct {
-			ContributionsCollection struct {
-				ContributionCalendar struct {
-					TotalContributions int
-					Weeks              []struct {
-						ContributionDays []struct {
-							ContributionCount int
-							Weekday           int
-							Date              string
-						}
-					}
-				}
-			}
-		} `graphql:"user(login: $username)"`
+type GetContributionsByUsernameOptions struct {
+	Username string
+	From time.Time
+	To time.Time
+}
+
+func (g *GithubService) GetContributionsByUsername(ctx context.Context, options GetContributionsByUsernameOptions) (*Contributions, error) {
+	var contributions *Contributions
+
+	if len(options.Username) == 0 {
+		return nil, ErrMissingUsername
 	}
 
-	err := g.githubClient.Query(ctx, &contributionsQuery, map[string]interface{}{
-		"username": githubv4.String(username),
-	})
-	if err != nil {
-		return nil, errors.Wrap(err, "github client query")
+	from, to := options.From, options.To
+
+	if to.IsZero() {
+		to = time.Now()
+	}
+
+	if from.IsZero() {
+		from = to.AddDate(-1, 0, 0)
 	}
 
-	contributions := &Contributions{
-		TotalContributions: contributionsQuery.User.ContributionsCollection.ContributionCalendar.TotalContributions,
+	years := date.GetYearsBetweenDates(from, to)
+
+	contributions = &Contributions{
+		TotalContributions: 			0,
 		Days:               []Day{},
 	}
 
-	for _, w := range contributionsQuery.User.ContributionsCollection.ContributionCalendar.Weeks {
+	if len(years) > 1 {
 
-		for _, d := range w.ContributionDays {
+		for _, year := range years {
 
-			date, err := time.Parse("2006-01-02", d.Date)
-			if err != nil {
-				return nil, errors.Wrap(err, "parsing date")
+			from = time.Date(year, 1, 1, 0, 0, 0, 0, time.UTC)
```

Comment:
I don't think this query works if it's a multi year query and the from date is in the middle of the year because we'd still go back to the start of the year since the from date is hard coded to the start of the year here

Reply:
example code above shows a more succinct way that also captures this edge case